File: Tree/count_complete_tree_nodes.rb
TreeNode = Struct.new(:val, :left, :right)

# @param {TreeNode} root
# @return {Integer}
def count_nodes(root)
    if root.nil?
        0
    else
        left_height = 1
        left = root.left
        while left
            left_height += 1
            left = left.left
        end

        right_height = 1
        right = root.right
        while right
            right_height += 1
            right = right.right
        end

        # puts "r: #{root.val} left_height: #{left_height}, right_height: #{right_height}"
        if left_height == right_height
            (2 ** right_height) - 1
        else
            1 + count_nodes(root.left) + count_nodes(root.right)
        end
    end
end

root = TreeNode.new(1)
root.left = TreeNode.new(2)
root.right = TreeNode.new(3)
root.left.left = TreeNode.new(4)
root.left.right = TreeNode.new(5)
root.right.left = TreeNode.new(6)
puts count_nodes(root) # 6

root.right.right = TreeNode.new(24)
puts count_nodes(root) # 7

File: leetcode/CountCompleteTreeNodes.java
/*
	https://leetcode.com/problems/count-complete-tree-nodes/
*/
class CountCompleteTreeNodes {
	public static void main(String[] args) {
		var root = new TreeNode(1);
		root.left = new TreeNode(2);
		root.right = new TreeNode(3);
		root.left.left = new TreeNode(5);
		root.left.right = new TreeNode(7);
		root.right.left = new TreeNode(6);

		System.out.println(new Solution().countNodes(root)); // 6

		System.out.println(new Solution().countNodes(null)); // 0

		System.out.println(new Solution().countNodes(new TreeNode(2))); // 1
	}
}

class Solution {
    public int countNodes(TreeNode root) {
        if (root == null) return 0;

        int height = findHeight(root);

        int left = 1;
        int right = (int)Math.pow(2, height);

        while(right - left > 1) {
            int mid = (left + right) / 2;

            if(isNodeExists(root, mid, height)) {
                left = mid;
            } else {
                right = mid - 1;
            }
        }

        int lastLeft = isNodeExists(root, right, height) ? right : left;

        return ((int)Math.pow(2, height)- 1) + lastLeft;
    }

    private boolean isNodeExists(TreeNode root, int num, int level) {
        if (root == null) return false;
        else if (level == 0) return true;

        int leftNodesCount = (int)Math.pow(2, level) / 2;

        if (leftNodesCount >= num) { // go left
            return isNodeExists(root.left, num, level - 1);
        } else { // go right
            return isNodeExists(root.right, num - leftNodesCount, level - 1);
        }
    }

    private int findHeight(TreeNode root) {
        if (root.left == null) return 0;
        return 1 + findHeight(root.left);
    }

}

/** Definition for a binary tree node. 	 */
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
}
